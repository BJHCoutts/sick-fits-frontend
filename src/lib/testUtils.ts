import casual from 'casual'
import { PAGINATION_QUERY } from './graphQL/queries/paginationQuery'
import { TProduct } from './types/TProduct'
import { TCartItem } from './types/TCartItem'
import { TUser } from './types/TUser'

// seed it so we get consistent results
casual.seed(777)

export const fakeItem = (): TProduct => ({
	__typename: 'Item',
	id: 'abc123',
	price: 5000,
	user: null,
	photo: {
		id: 'abc123',
		altText: 'dogs are best',
		image: {
			id: 'abc123',
			publicUrlTransformed: 'dog.jpg',
		},
	},
	name: 'dogs are best',
	description: 'dogs',
	status: 'AVAILABLE',
})

export const fakeUser = (overrides: {}): TUser => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
	...overrides,
})

// const fakeOrderItem = () => ({
//   __typename: 'OrderItem',
//   id: casual.uuid,
//   image: {
//     image: `${casual.word}.jpg`,
//   },
//   name: casual.words(),
//   price: 4234,
//   quantity: 1,
//   description: casual.words(),
// });

// const fakeOrder = () => ({
//   __typename: 'Order',
//   id: 'ord123',
//   charge: 'ch_123',
//   total: 40000,
//   items: [fakeOrderItem(), fakeOrderItem()],
//   createdAt: '2022-12-11T20:16:13.797Z',
//   user: fakeUser(),
// });

export const fakeCartItem = (overrides: {}): TCartItem => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 4,
	product: fakeItem(),
	user: fakeUser({}),
	...overrides,
})

// // Fake LocalStorage
// class LocalStorageMock {
//   constructor() {
//     this.store = {};
//   }

//   clear() {
//     this.store = {};
//   }

//   getItem(key) {
//     return this.store[key] || null;
//   }

//   setItem(key, value) {
//     this.store[key] = value.toString();
//   }

//   removeItem(key) {
//     delete this.store[key];
//   }
// }

export function makePaginationMocksFor(length: number) {
	return [
		{
			request: { query: PAGINATION_QUERY },
			result: {
				data: {
					_allProductsMeta: {
						count: length,
					},
					itemsConnection: {
						__typename: 'aggregate',
						aggregate: {
							count: length,
							__typename: 'count',
						},
					},
				},
			},
		},
	]
}

// export
// makePaginationMocksFor,
// LocalStorageMock,
// fakeItem,
// fakeUser,
// fakeCartItem,
// fakeOrder,
// fakeOrderItem,
