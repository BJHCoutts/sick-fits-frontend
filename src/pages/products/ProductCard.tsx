import Link from 'next/link'
import { TProduct } from '../../lib/types/TProduct'
import SItem from '../../components/styles/SItem'
import STitle from '../../components/styles/STitle'
import SPriceTag from '../../components/styles/SPriceTag'
import formatMoney from '../../lib/functions/formatMoney'

interface IProduct extends TProduct {}

export default function ProductCard({ product }: { product: IProduct }) {
	return (
		<Link href={`/products/${product.id}`}>
			<SItem>
				<img
					src={product?.photo?.image.publicUrlTransformed}
					alt={product?.photo?.altText}
				/>

				{!product.photo && <h3 className="no-photo">No Photo</h3>}

				<STitle>
					<Link href={`/product/${product.id}`}>{product.name}</Link>
				</STitle>
				<SPriceTag>{formatMoney(product.price)}</SPriceTag>
				<p>{product.description}</p>
				<div className="buttonlist">
					<Link
						href={{
							pathname: '/update',
							query: { id: product.id },
						}}
					>
						Edit üìù
					</Link>
				</div>
			</SItem>
		</Link>
	)
}
