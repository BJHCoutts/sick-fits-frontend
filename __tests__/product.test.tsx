import { render, screen } from '@testing-library/react'
import React from 'react'
import ProductCard from '../src/pages/products/ProductCard'
import { fakeProduct } from '../src/lib/testUtils'
import { MockedProvider } from '@apollo/client/testing'

describe('<ProductCard/>', () => {
	it('renders out the price and title', () => {
		const { container, debug } = render(
			<MockedProvider>
				<ProductCard product={fakeProduct} />
			</MockedProvider>
		)
		const priceTag = screen.getByText('$50')
		expect(priceTag).toBeInTheDocument()
		const link = container.querySelector('a')
		if (link) {
			expect(link).toHaveAttribute('href', '/product/abc123')
		}
		expect(container).toHaveTextContent(fakeProduct.name)
	})
})

it('Renders and matches HTML Snapshot', () => {
	const { container, debug } = render(
		<MockedProvider>
			<ProductCard product={fakeProduct} />
		</MockedProvider>
	)
	expect(container).toMatchSnapshot()
})

it('Renders the image properly', () => {
	const { container, debug } = render(
		<MockedProvider>
			<ProductCard product={fakeProduct} />
		</MockedProvider>
	)
	const img = screen.getByAltText(fakeProduct.name)
	expect(img).toBeInTheDocument()
})
