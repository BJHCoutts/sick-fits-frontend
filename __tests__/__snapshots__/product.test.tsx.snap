// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders and matches HTML Snapshot 1`] = `
<div>
  <li
    class="sc-bdfCDU ipTzGd"
  >
    <a
      href="/product/abc123"
    >
      <img
        alt="dogs are best"
        src="dog.jpg"
      />
    </a>
    <h3
      class="sc-gsTDqH dvgYyH"
    >
      <a
        href="/product/abc123"
      >
        dogs are best
      </a>
    </h3>
    <span
      class="sc-dlfmHC dNTxCS"
    >
      $50
    </span>
    <p>
      dogs
    </p>
    <div
      class="button-list"
    >
      <button
        aria-busy="false"
        type="button"
      >
        Add to Cart 🛒
      </button>
      <a
        href="/update?id=abc123"
      >
        Edit 📝
      </a>
      <button
        aria-busy="false"
        type="button"
      >
        Delete ☠️
      </button>
    </div>
  </li>
</div>
`;
